node/subsystem-util/src/tracked.rs

```diff
@@ -0,0 +1,252 @@
+}
+
+impl<T> TrackedReceiver<T> {
+	/// Count the number of items queued up inside the channel.
+	pub fn queue_count(&self) -> usize {
+		// when obtaining we don't care much about off by one
+		// accuracy
+		self.fill.load(Ordering::Relaxed)
+	}
+
+	/// Attempt to receive the next item.
+	pub fn try_next(&mut self) -> Result<Option<T>, mpsc::TryRecvError> {
+		match self.inner.try_next()? {
+			Some(x) => {
+				self.fill.fetch_sub(1, Ordering::SeqCst);
+				Ok(Some(x))
+			},
+			None => Ok(None),
+		}
+	}
+}
+
+/// The sender component, tracking the number of items
+/// sent across it.
+#[derive(Debug,Clone)]
+pub struct TrackedSender<T> {
+	fill: Arc<AtomicUsize>,
+	inner: mpsc::Sender<T>,
+	name: &'static str,
+}
+
+impl<T> std::ops::Deref for TrackedSender<T> {
+	type Target = mpsc::Sender<T>;
+	fn deref(&self) -> &Self::Target {
+		&self.inner
+	}
+}
+
+impl<T> std::ops::DerefMut for TrackedSender<T> {
+	fn deref_mut(&mut self) -> &mut Self::Target {
+		&mut self.inner
+	}
+}
+
+impl<T> TrackedSender<T> {
+	/// Count the number of items already queued up for consumption on
+	/// on the rx side.
+	pub fn queue_count(&self) -> usize {
+		self.fill.load(Ordering::Relaxed)
+	}
+
+	/// Send message, wait until capacity is available.
+	pub async fn send(&mut self, item: T) -> result::Result<(), mpsc::SendError> where Self: Unpin {
+		self.fill.fetch_add(1, Ordering::SeqCst);
+		let fut = self.inner.send(item);
+		futures::pin_mut!(fut);
+		fut.await
+	}
+
+	/// Attempt to send message or fail immediately.
+	pub fn try_send(&mut self, msg: T) -> result::Result<(), mpsc::TrySendError<T>> {
+		self.inner.try_send(msg)?;
+		self.fill.fetch_add(1, Ordering::SeqCst);
+		Ok(())
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use futures::executor::block_on;
+	use futures::StreamExt;
+
+	#[derive(Clone, Copy, Debug, Default)]
+	struct Msg {
+		val: u8,
+	}
+
+	#[test]
+	fn try_send_try_next() {
+		block_on(async move {
+			let (mut tx, mut rx) = channel::<Msg>(5, "goofy");
+			let msg = Msg::default();
+			assert_eq!(rx.queue_count(), 0);
+			tx.try_send(msg).unwrap();
+			assert_eq!(tx.queue_count(), 1);
+			tx.try_send(msg).unwrap();
+			tx.try_send(msg).unwrap();
+			tx.try_send(msg).unwrap();
+			assert_eq!(tx.queue_count(), 4);
+			rx.try_next().unwrap();
+			assert_eq!(rx.queue_count(), 3);
+			rx.try_next().unwrap();
+			rx.try_next().unwrap();
+			assert_eq!(tx.queue_count(), 1);
+			rx.try_next().unwrap();
+			assert_eq!(rx.queue_count(), 0);
+			assert!(rx.try_next().is_err());
+		});
+	}
+
+	#[test]
+	fn with_tasks() {
+		let (mut tx, mut rx) = channel::<Msg>(5, "goofy");
+		block_on(async move {
+			futures::join!(
+				async move {
+					let msg = Msg::default();
+					assert_eq!(tx.queue_count(), 0);
+					tx.try_send(msg).unwrap();
+					assert_eq!(tx.queue_count(), 1);
+					tx.try_send(msg).unwrap();
+					tx.try_send(msg).unwrap();
+					tx.try_send(msg).unwrap();
+				},
+				async move {
+					assert_eq!(rx.queue_count(), 4);
```

Comment:
I don't think that this test can work properly unless you have some kind of external synchronization which ensures that we don't attempt this assertion until the other future has run to completion.

Reply:
Thanks, mere luck this did not blow up.